core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala

```diff
@@ -0,0 +1,648 @@
+   * @return a string value, primarily for testing and diagnostics
+   */
+  override def toString: String = {
+    val sb = new StringBuilder(s"ApplicationCache(refreshInterval=$refreshInterval," +
+          s" retainedApplications= $retainedApplications)")
+    sb.append(s"; time= ${clock.getTimeMillis()}")
+    sb.append(s"; entry count= ${appCache.size()}\n")
+    sb.append("----\n")
+    appCache.asMap().asScala.foreach {
+      case(key, entry) => sb.append(s"  $key -> $entry\n")
+    }
+    sb.append("----\n")
+    sb.append(metrics)
+    sb.append("----\n")
+    sb.toString()
+  }
+}
+
+/**
+ * An entry in the cache.
+ *
+ * @param ui Spark UI
+ * @param completed: flag to indicated that the application has completed (and so
+ *                 does not need refreshing)
+ * @param probeTime timestamp in milliseconds. This may be updated during probes
+ */
+private[history] final class CacheEntry(val ui: SparkUI, val completed: Boolean,
+    val data: Option[HistoryProviderUpdateState],
+    val loadTime: Long,
+    var probeTime: Long) {
+
+  /** string value is for test assertions */
+  override def toString: String = {
+    s"UI $ui, data=$data, completed=$completed, loadTime=$loadTime probeTime=$probeTime"
+  }
+}
+
+/**
+ * Cache key: compares on App Id and then, if non-empty, attemptId.
+ * The [[hashCode()]] function uses the same fields.
+ * @param appId application ID
+ * @param attemptId attempt ID
+ */
+private[history] final case class CacheKey(appId: String, attemptId: Option[String]) {
+
+  override def hashCode(): Int = {
+    appId.hashCode() + attemptId.map(_.hashCode).getOrElse(0)
+  }
+
+  override def equals(obj: Any): Boolean = {
+    val that = obj.asInstanceOf[CacheKey]
+    that.appId == appId && that.attemptId == attemptId
+  }
+
+  override def toString: String = {
+    appId + attemptId.map { id => s"/$id" }.getOrElse("")
+  }
+}
+
+private[history] class CacheMetrics(prefix: String) extends Source {
+
+  /* metrics: counters and timers */
+  val lookupCount = new Counter()
+  val lookupFailureCount = new Counter()
+  val evictionCount = new Counter()
+  val loadCount = new Counter()
+  val loadTimer = new Timer()
+  val updateProbeCount = new Counter()
+  val updateProbeTimer = new Timer()
+  val updateTriggeredCount = new Counter()
+
+  /** all the counters: for registration and string conversion. */
+  private val counters = Seq(
+    ("lookup.count", lookupCount),
+    ("lookup.failure.count", lookupFailureCount),
+    ("eviction.count", evictionCount),
+    ("load.count", loadCount),
+    ("update.probe.count", updateProbeCount),
+    ("update.triggered.count", updateTriggeredCount))
+
+  /** all metrics, including timers */
+  private val allMetrics = counters ++ Seq(
+    ("load.timer", loadTimer),
+    ("update.probe.timer", updateProbeTimer))
+
+  /**
+   * Name of metric source
+   */
+  override val sourceName = "ApplicationCache"
+
+  override def metricRegistry: MetricRegistry = new MetricRegistry
```

Comment:
I think you want to store this in a `val` and not return a new one on every call